app/src/main/java/com/example/sharingang/UserProfileFragment.kt

```diff
@@ -59,67 +90,89 @@ class UserProfileFragment : Fragment() {
             null, "" -> currentUserId
             else -> args.userId
         }
+        authHelper = AuthHelper(resultLauncher, lifecycleScope, currentUser, userRepository)
+        currentUser = currentUserProvider.getCurrentUser()
+        setUserType()
         userViewModel.setUser(shownUserProfileId)
         imageAccess = ImageAccess(requireActivity())
         imageAccess.setupImageView(binding.imageView)
         lifecycle.addObserver(imageAccess)
-        userViewModel.user.observe(viewLifecycleOwner, { user ->
-            displayUserFields(user)
-        })
-        setupRecyclerView(shownUserProfileId)
         binding.viewModel = userViewModel
+        initialize_fields()
+        return binding.root
+    }
+
+    private fun initialize_fields() {
+        currentUserId = currentUserProvider.getCurrentUserId()
+        setUserType()
+        setupRecyclerView(shownUserProfileId)
         loggedInUserEmail = currentUserProvider.getCurrentUserEmail()
+        setupButtons()
+        authHelper.restoreLoginStatus()
         initSetup()
-        setupViewAndButtonsAction()
+        setEmailText()
+        setVisibilities()
+        setupViews()
         setupReportButton()
         setupRatingView()
-        return binding.root
+        userViewModel.user.observe(viewLifecycleOwner, { user ->
+            displayUserFields(user)
+        })
+    }
+
+    private fun setUserType() {
+        userType = when (currentUserId) {
+            null -> UserType.LOGGED_OUT
+            shownUserProfileId -> UserType.SELF
+            else -> UserType.VISITOR
+        }
+        if(userType == UserType.LOGGED_OUT && shownUserProfileId == null) {
+            userType = UserType.LOGGED_OUT_SELF
+        }
     }
 
     private fun initSetup() {
-        val fields = listOf(
-            binding.imageView,
-            binding.gallerycameraholder,
-            binding.nameText,
-            binding.textEmail,
-            binding.applyholder,
-            binding.ratingTextview,
-            binding.applyholder,
-            binding.btnReport
-        )
-        for (view: View in fields) {
-            view.visibility = View.GONE
+        listOf(binding.upfTopinfo, binding.imageView, binding.gallerycameraholder, binding.nameText,
+            binding.textEmail, binding.applyholder, binding.ratingTextview, binding.applyholder,
+            binding.btnReport, binding.userItemList, binding.btnLogout,
+            binding.btnLogin
+        ).forEach { view -> view.visibility = View.GONE }
```

Comment:
Do you know a straightforwards way to do that? What I was thinking is setting them all to GONE initially (i.e. in `user_profile_fragment.xml`), but then it would be a pain to modify the layout if we want to add new buttons etc. I thought this was fine but if there is a better way to do that (i.e. getting the list/set of all views in the fragment) that would indeed be better.

Reply:
Unfortunately I don't know of a way to do that, so let's keep it like this